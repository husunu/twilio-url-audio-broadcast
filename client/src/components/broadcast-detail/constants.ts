import { Broadcast, Channel } from '@/types';

export const defaultBroadcast: Broadcast = {
  channel: Channel.VOICE,
  friendlyName: '',
  from: '',
  broadcastId: '',
  message: '',
  dateCreated: new Date(),
  canceled: false,
  completed: false,
};

export const LOADER_TEST_ID = 'broadcast-detail-loader';
export const RECIPIENTS_TABLE_TEST_ID = 'recipients-table';
export const PAGINATOR_TEST_ID = 'paginator';
