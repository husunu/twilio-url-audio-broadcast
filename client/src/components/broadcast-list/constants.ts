export const LOADER_TEST_ID = 'broadcast-list-loader';
export const REQUIREMENTS = [
  'Active Twilio Phone number with Voice capabilities',
  'List of recipient phone numbers (.txt)',
  'Message you want to send',
];
