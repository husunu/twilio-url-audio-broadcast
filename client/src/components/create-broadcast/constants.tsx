import React from 'react';

import { StepItem } from './types';
import { StepType } from '@/types';

import { Configure, RecipientList, Review } from './steps';

import {
  Settings as SettingsIcon,
  GroupAdd as GroupAddIcon,
  Send as SendIcon,
} from '@material-ui/icons';

const locations = {
  configure: '/create/configure',
  recipients: '/create/recipients',
  review: '/create/review',
};

export const steps: Array<StepItem> = [
  {
    key: StepType.Configure,
    label: 'Configure Notification',
    icon: <SettingsIcon />,
    location: locations.configure,
    component: Configure,
    next: locations.recipients,
  },
  {
    key: StepType.Recipients,
    label: 'Upload Recipients',
    icon: <GroupAddIcon />,
    location: locations.recipients,
    component: RecipientList,
    prev: locations.configure,
    next: locations.review,
  },
  {
    key: StepType.Review,
    label: 'Review & Send',
    icon: <SendIcon />,
    location: locations.review,
    component: Review,
    prev: locations.recipients,
    next: '',
  },
];
