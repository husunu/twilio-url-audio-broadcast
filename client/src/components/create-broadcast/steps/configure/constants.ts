export const buyNewNumberUrl =
  'https://www.twilio.com/console/phone-numbers/search';
