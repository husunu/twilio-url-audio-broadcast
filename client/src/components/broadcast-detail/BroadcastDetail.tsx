import React, { useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';
import { useBroadcast } from './hooks/useBroadcast';
import { useRecipients } from './hooks/useRecipients';
import { useCancelBroadcast } from './hooks/useCancelBroadcast';
import { StyledCard } from '@/components/common/styled-card';
import { defaultBroadcast } from './constants';
import { LOADER_TEST_ID } from './constants';
import { Alert, AlertType } from '@/components/common/alert';
import { CallStatusTable } from './CallStatusTable';
import { AlertArea } from './AlertArea';
import { NotificationBasicinfo } from './NotificationBasicInfo';
import { Summary } from './Summary';
import { NotificationExtendedinfo } from './NotificationExtendedinfo';
import { BroadcastDetailNavigation } from './BroadcastDetailNavigation';
import { Box, CircularProgress, Grid, Typography } from '@material-ui/core';

export const BroadcastDetail = () => {
  const location = useLocation();

  const [showSuccess, setShowSuccess] = useState<boolean>(false);
  const [currentPage, setCurrentPage] = useState<number>(0);
  const [rowsPerPage, setRowsPerPage] = useState<number>(10);
  const [hasDownloadReportFailed, setHasDownloadReporFailed] = useState<
    boolean
  >(false);

  const {
    data,
    loading: broadcastLoading,
    error: broadcastError,
    get: getBroadcast,
  } = useBroadcast();

  const { broadcast = defaultBroadcast } = data;

  const {
    recipients,
    meta,
    pageCount,
    loading: recipientsLoading,
    error: recipientsError,
  } = useRecipients(currentPage, rowsPerPage);

  const {
    patch: cancelBroadcast,
    response: cancelBroadcastResponse,
    loading: cancelBroadcastLoading,
    error: cancelBroadcastError,
  } = useCancelBroadcast();

  useEffect(() => {
    const params = new URLSearchParams(location.search);
    const isSuccess = params.get('success');
    if (isSuccess) {
      setShowSuccess(true);
    }
  }, [location, setShowSuccess]);

  if (broadcastLoading) {
    return (
      <Box data-testid={LOADER_TEST_ID}>
        <CircularProgress />
      </Box>
    );
  }

  if (broadcastError) {
    return (
      <Alert
        alert={{
          message:
            'Sorry, something failed and we could not cancel the notification.',
          type: AlertType.ERROR,
        }}
        showReload
      />
    );
  }

  return (
    <>
      <BroadcastDetailNavigation
        broadcast={broadcast}
        meta={meta}
        cancelBroadcast={cancelBroadcast}
        cancelBroadcastLoading={cancelBroadcastLoading}
        cancelBroadcastResponse={cancelBroadcastResponse}
        getBroadcast={getBroadcast}
        setHasDownloadReporFailed={setHasDownloadReporFailed}
      />
      <Box my={3}>
        <StyledCard>
          <AlertArea
            showSuccess={showSuccess}
            recipientsError={recipientsError}
            hasDownloadReportFailed={hasDownloadReportFailed}
            cancelBroadcastError={cancelBroadcastError}
          />

          <Box px={3} pt={3}>
            <Typography variant="h4" component="h1">
              Notification Report
            </Typography>
          </Box>

          {!broadcastLoading && (
            <Grid container>
              {/* LEFT COLUMN */}
              <Grid item xs={6}>
                <Box p={3}>
                  <NotificationBasicinfo broadcast={broadcast} meta={meta} />
                  {!recipientsError && meta && meta.count && (
                    <Summary meta={meta} />
                  )}
                </Box>
              </Grid>

              {/* RIGHT COLUMN */}
              <Grid item xs={6}>
                <Box p={3}>
                  <NotificationExtendedinfo broadcast={broadcast} meta={meta} />
                  <CallStatusTable
                    recipients={recipients}
                    meta={meta}
                    loading={recipientsLoading}
                    pageCount={pageCount}
                    rowsPerPage={rowsPerPage}
                    setRowsPerPage={setRowsPerPage}
                    currentPage={currentPage}
                    setCurrentPage={setCurrentPage}
                  />
                </Box>
              </Grid>
            </Grid>
          )}
        </StyledCard>
      </Box>
    </>
  );
};
