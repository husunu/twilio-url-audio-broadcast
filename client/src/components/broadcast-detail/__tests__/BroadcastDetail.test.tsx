import React from 'react';
import { cleanup, waitFor } from '@testing-library/react';
import { FetchMock } from 'jest-fetch-mock';
import MockDate from 'mockdate';
import moment from 'moment-timezone';

import { BroadcastDetail } from '../BroadcastDetail';
import { LOADER_TEST_ID, RECIPIENTS_TABLE_TEST_ID } from '../constants';
import { renderWithAppContexts } from '@/testUtils';
import { act } from 'react-dom/test-utils';

const fetch = global.fetch as FetchMock;
moment.tz.setDefault('America/Los_Angeles');

const sampleBroadcastResponse = {
  broadcast: {
    broadcastId: 'd4600746-2e8d-40cf-8a84-036cb20d9db7',
    friendlyName: 'Broadcast sample',
    message: 'Ahoy there!',
    from: '+12023187895',
    channel: 'voice',
    dateCreated: moment('2020-06-03 15:28:04').toISOString(),
    canceled: false,
  },
};

const sampleRecipientsResponse = {
  recipients: [
    {
      callSid: 'CA3f8433dc395551977d79d84bee54bf04',
      to: '+12015650934',
      status: 'completed',
      broadcastId: 'd4600746-2e8d-40cf-8a84-036cb20d9db7',
    },
    {
      callSid: 'CA73afe9a827db1bc18a14881b3b232ddb',
      to: '+12015723115',
      status: 'completed',
      broadcastId: 'd4600746-2e8d-40cf-8a84-036cb20d9db7',
    },
    {
      callSid: 'CA2eeb204c1fa1b898cb03fc7126d4ebff',
      to: '+12015711563',
      status: 'busy',
      broadcastId: 'd4600746-2e8d-40cf-8a84-036cb20d9db7',
    },
    {
      callSid: 'CAe5ebee100422399015e1c8a38ddc6413',
      to: '+12015552572',
      status: 'completed',
      broadcastId: 'd4600746-2e8d-40cf-8a84-036cb20d9db7',
    },
    {
      callSid: 'CA45bf025dd72af2a7ebeb3c7d37c8087a',
      to: '+12015581526',
      status: 'in-progress',
      broadcastId: 'd4600746-2e8d-40cf-8a84-036cb20d9db7',
    },
    {
      callSid: 'CAc04cb434c1434dbf8c3f0a479501ce1b',
      to: '+12015553700',
      status: 'in-progress',
      broadcastId: 'd4600746-2e8d-40cf-8a84-036cb20d9db7',
    },
    {
      callSid: 'CAc8900bfd65037303c904188d7edbaf3c',
      to: '+12015667027',
      status: 'ringing',
      broadcastId: 'd4600746-2e8d-40cf-8a84-036cb20d9db7',
    },
    {
      callSid: 'CAd30a3795ec713def1a819c2defc0fe38',
      to: '+12015623466',
      status: 'queued',
      broadcastId: 'd4600746-2e8d-40cf-8a84-036cb20d9db7',
    },
    {
      callSid: 'CA02be6e98a0cb2106fe1bcc76aa6f5db6',
      to: '+12015598376',
      status: 'queued',
      broadcastId: 'd4600746-2e8d-40cf-8a84-036cb20d9db7',
    },
    {
      callSid: 'CAdfe31eab845e8daa4220a16a20a1e4af',
      to: '+12015612792',
      status: 'queued',
      broadcastId: 'd4600746-2e8d-40cf-8a84-036cb20d9db7',
    },
  ],
  pageCount: 2,
  meta: {
    total: 20,
    completed: 4,
    count: {
      busy: 1,
      canceled: 0,
      completed: 2,
      failed: 1,
      'in-progress': 0,
      initiated: 0,
      'no-answer': 0,
      queued: 14,
      ringing: 2,
    },
    broadcastStatus: 'in-progress',
  },
};

describe('BroadcastDetail', () => {
  beforeAll(() => {
    MockDate.set(1592838350131); // Mon Jun 22 2020 17:05:50 GMT+0200 (Central European Summer Time)
  });

  afterEach(() => {
    cleanup();
    fetch.resetMocks();
  });

  afterAll(() => {
    MockDate.reset();
  });

  test('loader is shown while loading', async () => {
    await act(async () => {
      jest.useFakeTimers();
      fetch.mockResponse(
        () =>
          new Promise((resolve) =>
            setTimeout(() => resolve({ body: 'ok' }), 200)
          )
      );
      const { getByTestId } = renderWithAppContexts(<BroadcastDetail />);
      expect(getByTestId(LOADER_TEST_ID)).toBeInTheDocument();
      jest.clearAllTimers();
    });
  });

  test('data is rendered correctly', async () => {
    fetch.mockResponses(
      [JSON.stringify(sampleBroadcastResponse), { status: 200 }],
      [JSON.stringify(sampleRecipientsResponse), { status: 200 }],
      [JSON.stringify(sampleRecipientsResponse), { status: 200 }]
    );

    const { container, getByTestId } = renderWithAppContexts(
      <BroadcastDetail />
    );

    await waitFor(() => {
      expect(getByTestId(RECIPIENTS_TABLE_TEST_ID)).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });

  test('broadcast request error is handled correctly', async () => {
    fetch.mockReject(new Error('Internal Server Error'));

    const { container, getByText } = renderWithAppContexts(<BroadcastDetail />);
    await waitFor(() => {
      expect(
        getByText(
          'Sorry, something failed and we could not cancel the notification.'
        )
      ).toBeInTheDocument();

      expect(container).toMatchSnapshot();
    });
  });

  test('broadcast recipients request error is handled correctly', async () => {
    fetch.mockResponses(
      [JSON.stringify(sampleBroadcastResponse), { status: 200 }],
      [JSON.stringify(new Error('Internal Server Error')), { status: 500 }],
      [JSON.stringify(new Error('Internal Server Error')), { status: 500 }]
    );

    const { container, getByText } = renderWithAppContexts(<BroadcastDetail />);
    await waitFor(() => {
      expect(
        getByText(
          'Sorry, we could not retrieve your notification data this time.'
        )
      ).toBeInTheDocument();

      expect(container).toMatchSnapshot();
    });
  });
});
