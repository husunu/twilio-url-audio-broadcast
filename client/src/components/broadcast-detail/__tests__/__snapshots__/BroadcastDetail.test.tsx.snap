// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BroadcastDetail broadcast recipients request error is handled correctly 1`] = `
<div>
  <nav
    class="MuiGrid-root MuiGrid-container"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <a
        aria-disabled="false"
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
        href="/broadcasts"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiButton-label"
        >
          Back
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </a>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      style="text-align: right;"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Cancel notification
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </nav>
  <div
    class="MuiBox-root MuiBox-root-654"
  >
    <div
      class="MuiPaper-root MuiCard-root WithStyles(ForwardRef(Card))-root-655 WithStyles(ForwardRef(Card))-root-656 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiBox-root MuiBox-root-736"
      >
        <div
          class="MuiPaper-root MuiAlert-root MuiAlert-standardError MuiPaper-elevation0"
          role="alert"
          style="align-items: center;"
        >
          <div
            class="MuiAlert-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
          </div>
          <div
            class="MuiAlert-message"
          >
            Sorry, we could not retrieve your notification data this time.
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root MuiBox-root-686"
      >
        <h1
          class="MuiTypography-root MuiTypography-h4"
        >
          Notification Report
        </h1>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <div
            class="MuiBox-root MuiBox-root-717"
          >
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Notification Name
            </h2>
            <div
              class="MuiBox-root MuiBox-root-718"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1"
              >
                Broadcast sample
              </h6>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              From Number
            </h2>
            <div
              class="MuiBox-root MuiBox-root-719"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1"
              >
                +12023187895
              </h6>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Notification Message
            </h2>
            <div
              class="MuiBox-root MuiBox-root-720"
            >
              <p
                class="MuiTypography-root MuiTypography-body2"
              >
                Ahoy there!
              </p>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Notification Status
            </h2>
            <div
              class="MuiBox-root MuiBox-root-721"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1"
              >
                -
              </h6>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <div
            class="MuiBox-root MuiBox-root-722"
          >
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Start time
            </h2>
            <div
              class="MuiBox-root MuiBox-root-723"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1"
              >
                6/3/2020, 3:28:04 PM
              </h6>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              End time
            </h2>
            <div
              class="MuiBox-root MuiBox-root-724"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1"
              >
                6/22/2020, 8:05:50 AM
              </h6>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Recipients
            </h2>
            <div
              class="MuiBox-root MuiBox-root-725"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1"
              >
                 Total Recipients
              </h6>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BroadcastDetail broadcast request error is handled correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiAlert-root MuiAlert-standardError MuiPaper-elevation0"
    role="alert"
    style="align-items: center;"
  >
    <div
      class="MuiAlert-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <div
      class="MuiAlert-message"
    >
      Sorry, something failed and we could not cancel the notification.
    </div>
    <div
      class="MuiAlert-action"
    >
      <div
        class="MuiBox-root MuiBox-root-447"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Reload
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`BroadcastDetail data is rendered correctly 1`] = `
<div>
  <nav
    class="MuiGrid-root MuiGrid-container"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <a
        aria-disabled="false"
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
        href="/broadcasts"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiButton-label"
        >
          Back
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </a>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      style="text-align: right;"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Cancel notification
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </nav>
  <div
    class="MuiBox-root MuiBox-root-179"
  >
    <div
      class="MuiPaper-root MuiCard-root WithStyles(ForwardRef(Card))-root-180 WithStyles(ForwardRef(Card))-root-181 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiBox-root MuiBox-root-211"
      >
        <h1
          class="MuiTypography-root MuiTypography-h4"
        >
          Notification Report
        </h1>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <div
            class="MuiBox-root MuiBox-root-242"
          >
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Notification Name
            </h2>
            <div
              class="MuiBox-root MuiBox-root-243"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1"
              >
                Broadcast sample
              </h6>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              From Number
            </h2>
            <div
              class="MuiBox-root MuiBox-root-244"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1"
              >
                +12023187895
              </h6>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Notification Message
            </h2>
            <div
              class="MuiBox-root MuiBox-root-245"
            >
              <p
                class="MuiTypography-root MuiTypography-body2"
              >
                Ahoy there!
              </p>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Notification Status
            </h2>
            <div
              class="MuiBox-root MuiBox-root-246"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1"
              >
                In progress (20%)
              </h6>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Summary
            </h2>
            <div
              class="MuiBox-root MuiBox-root-304"
            >
              <div
                class="MuiPaper-root MuiAlert-root MuiAlert-standardInfo MuiPaper-elevation0"
                role="alert"
              >
                <div
                  class="MuiAlert-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiAlert-message"
                >
                  This is a summary of the calls results used to send your voice notification. To learn more about call status and definitions visit
                   
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                    href="https://www.twilio.com/docs/voice/api/call-resource#call-status-values"
                    rel="noopener"
                    target="_blank"
                  >
                    Twilio Call Resource documentation
                  </a>
                </div>
              </div>
            </div>
            <ul
              class="MuiList-root MuiList-dense MuiList-padding"
            >
              <li
                class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
                style="padding: 0px;"
              >
                <div
                  class="MuiListItemText-root MuiListItemText-dense"
                  style="margin: 0px;"
                >
                  <div
                    class="MuiTypography-root MuiTypography-subtitle1"
                  >
                    14 - queued
                  </div>
                </div>
              </li>
              <li
                class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
                style="padding: 0px;"
              >
                <div
                  class="MuiListItemText-root MuiListItemText-dense"
                  style="margin: 0px;"
                >
                  <div
                    class="MuiTypography-root MuiTypography-subtitle1"
                  >
                    0 - initiated
                  </div>
                </div>
              </li>
              <li
                class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
                style="padding: 0px;"
              >
                <div
                  class="MuiListItemText-root MuiListItemText-dense"
                  style="margin: 0px;"
                >
                  <div
                    class="MuiTypography-root MuiTypography-subtitle1"
                  >
                    2 - ringing
                  </div>
                </div>
              </li>
              <li
                class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
                style="padding: 0px;"
              >
                <div
                  class="MuiListItemText-root MuiListItemText-dense"
                  style="margin: 0px;"
                >
                  <div
                    class="MuiTypography-root MuiTypography-subtitle1"
                  >
                    0 - in-progress
                  </div>
                </div>
              </li>
              <li
                class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
                style="padding: 0px;"
              >
                <div
                  class="MuiListItemText-root MuiListItemText-dense"
                  style="margin: 0px;"
                >
                  <div
                    class="MuiTypography-root MuiTypography-subtitle1"
                  >
                    2 - completed
                  </div>
                </div>
              </li>
              <li
                class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
                style="padding: 0px;"
              >
                <div
                  class="MuiListItemText-root MuiListItemText-dense"
                  style="margin: 0px;"
                >
                  <div
                    class="MuiTypography-root MuiTypography-subtitle1"
                  >
                    1 - busy
                  </div>
                </div>
              </li>
              <li
                class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
                style="padding: 0px;"
              >
                <div
                  class="MuiListItemText-root MuiListItemText-dense"
                  style="margin: 0px;"
                >
                  <div
                    class="MuiTypography-root MuiTypography-subtitle1"
                  >
                    0 - no-answer
                  </div>
                </div>
              </li>
              <li
                class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
                style="padding: 0px;"
              >
                <div
                  class="MuiListItemText-root MuiListItemText-dense"
                  style="margin: 0px;"
                >
                  <div
                    class="MuiTypography-root MuiTypography-subtitle1"
                  >
                    0 - canceled
                  </div>
                </div>
              </li>
              <li
                class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
                style="padding: 0px;"
              >
                <div
                  class="MuiListItemText-root MuiListItemText-dense"
                  style="margin: 0px;"
                >
                  <div
                    class="MuiTypography-root MuiTypography-subtitle1"
                  >
                    1 - failed
                  </div>
                </div>
              </li>
            </ul>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <div
            class="MuiBox-root MuiBox-root-247"
          >
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Start time
            </h2>
            <div
              class="MuiBox-root MuiBox-root-248"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1"
              >
                6/3/2020, 3:28:04 PM
              </h6>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              End time
            </h2>
            <div
              class="MuiBox-root MuiBox-root-249"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1"
              >
                -
              </h6>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Recipients
            </h2>
            <div
              class="MuiBox-root MuiBox-root-250"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1"
              >
                20
                 Total Recipients
              </h6>
            </div>
            <div
              class="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
              style="margin-bottom: 4px;"
            >
              <table
                aria-label="Notification recipients calls details and statuses"
                class="MuiTable-root"
                data-testid="recipients-table"
              >
                <thead
                  class="MuiTableHead-root"
                >
                  <tr
                    class="MuiTableRow-root MuiTableRow-head"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                      scope="col"
                    >
                      Number
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                      scope="col"
                    >
                      Status
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="MuiTableBody-root"
                >
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      +12015650934
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      completed
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      +12015723115
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      completed
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      +12015711563
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      busy
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      +12015552572
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      completed
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      +12015581526
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      in-progress
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      +12015553700
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      in-progress
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      +12015667027
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      ringing
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      +12015623466
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      queued
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      +12015598376
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      queued
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      +12015612792
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      queued
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div
              class="MuiTablePagination-root"
              data-testid="paginator"
            >
              <div
                class="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
              >
                <div
                  class="MuiTablePagination-spacer"
                />
                <p
                  class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
                >
                  1-10 of 20
                </p>
                <div
                  class="MuiTablePagination-actions"
                >
                  <button
                    aria-label="Previous page"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
                    disabled=""
                    tabindex="-1"
                    title="Previous page"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                        />
                      </svg>
                    </span>
                  </button>
                  <button
                    aria-label="Next page"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
                    tabindex="0"
                    title="Next page"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
